crates/accelerate/src/synthesis/multi_controlled/mcmt.rs

```diff
@@ -0,0 +1,181 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use core::num;
+
+use pyo3::prelude::*;
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::circuit_instruction::OperationFromPython;
+use qiskit_circuit::operations::{Param, StandardGate};
+use qiskit_circuit::packed_instruction::PackedOperation;
+use qiskit_circuit::{Clbit, Qubit};
+use smallvec::{smallvec, SmallVec};
+
+/// A Toffoli chain, implementing a multi-control condition on all controls using
+/// ``controls.len() - 1`` auxiliary qubits.
+///
+/// For example, for 4 controls we require 3 auxiliaries and create the circuit
+///
+///     control_0: ──■──────────────
+///                  │              
+///     control_1: ──■──────────────
+///                  │              
+///     control_2: ──┼────■─────────
+///                  │    │         
+///     control_3: ──┼────┼────■────
+///                ┌─┴─┐  │    │    
+///         aux_0: ┤ X ├──■────┼────
+///                └───┘┌─┴─┐  │    
+///         aux_1: ─────┤ X ├──■────
+///                     └───┘┌─┴─┐         "master control" qubit: controlling on this
+///         aux_2: ──────────┤ X ├──  <--  implements a controlled operation on all qubits
+///                          └───┘         in the "control" register
+fn ccx_chain<'a>(
+    controls: &'a [usize],
+    auxiliaries: &'a [usize],
+) -> impl DoubleEndedIterator<
+    Item = (
+        PackedOperation,
+        SmallVec<[Param; 3]>,
+        Vec<Qubit>,
+        Vec<Clbit>,
+    ),
+> + 'a {
+    let n = controls.len() - 1; // number of chain elements
+    let indices = std::iter::once((controls[0], controls[1], auxiliaries[0]))
+        .chain((0..n - 1).map(|i| (controls[i + 2], auxiliaries[i], auxiliaries[i + 1])));
+    indices.map(|(ctrl1, ctrl2, target)| {
```

Comment:
Do we even need to create the variable `indices`, i.e. have something like this `std::iter::once(...).chain(...).map(...)`?